Armadillo/src/main/java/com/scribd/armadillo/download/drm/ArmadilloDrmSessionManagerProvider.kt

```diff
@@ -0,0 +1,136 @@
+    override fun get(mediaItem: MediaItem): DrmSessionManager {
+        Assertions.checkNotNull(mediaItem.localConfiguration)
+        val drmConfiguration = mediaItem.localConfiguration!!.drmConfiguration
+        if (drmConfiguration != null && Util.SDK_INT >= 18) {
+            synchronized(this.lock) {
+                if (!Util.areEqual(drmConfiguration, this.drmConfiguration)) {
+                    this.drmConfiguration = drmConfiguration
+                    this.manager = this.createManager(drmConfiguration)
+                }
+                return Assertions.checkNotNull<DrmSessionManager?>(this.manager)
+            }
+        } else {
+            return DrmSessionManager.DRM_UNSUPPORTED
+        }
+    }
+
+    /** Near identical to DefaultDrmSessionManagerProvider method, except for the indicated line */
+    private fun createManager(drmConfiguration: DrmConfiguration): DrmSessionManager {
+        val dataSourceFactory = this.drmHttpDataSourceFactory
+            ?: DefaultHttpDataSource.Factory().setUserAgent(this.userAgent)
+        val httpDrmCallback = HttpMediaDrmCallback(if (drmConfiguration.licenseUri == null) null else drmConfiguration.licenseUri.toString(), drmConfiguration.forceDefaultLicenseUri, (dataSourceFactory)!!)
+        val var4: UnmodifiableIterator<*> = drmConfiguration.licenseRequestHeaders.entries.iterator()
+
+        while (var4.hasNext()) {
+            val entry: Map.Entry<String, String> = var4.next() as Map.Entry<String, String>
+            httpDrmCallback.setKeyRequestProperty(entry.key, entry.value)
+        }
+
+        val drmSessionManager = DefaultDrmSessionManager.Builder()
+            //this line is different, changing the DrmProvider
+            .setUuidAndExoMediaDrmProvider(drmConfiguration.scheme, ArmadilloDrmProvider(stateStore))
+            .setMultiSession(drmConfiguration.multiSession)
+            .setPlayClearSamplesWithoutKeys(drmConfiguration.playClearContentWithoutKey)
+            .setUseDrmSessionsForClearContent(*Ints.toArray(drmConfiguration.forcedSessionTrackTypes))
+            .build(httpDrmCallback)
+        drmSessionManager.setMode(DefaultDrmSessionManager.MODE_PLAYBACK, drmConfiguration.keySetId)
+        return drmSessionManager
+    }
+
+    /** New original provider for this class */
+    class ArmadilloDrmProvider(private val stateStore: StateStore.Modifier) : ExoMediaDrm.Provider {
+        private var drmExpirationMillis: Long? = null
+        override fun acquireExoMediaDrm(uuid: UUID): ExoMediaDrm {
+            Log.i(TAG, "drm start")
+            //uses the main MediaDrm object that this class uses originally, then after we add new listeners to it.
+            val instance = FrameworkMediaDrm.newInstance(uuid)
+
+            //ExoMediaDrm.OnEventListener doesn't do anything at all, so its not used
+            if (Util.SDK_INT >= 23) {
+                instance.setOnKeyStatusChangeListener { exoMediaDrm, sessionId, keyStatuses, b ->
```

Comment:
If we're gonna keep all the params names (are we using them later?), the last boolean param is actually `hasNewUsableKey` 😁